pkg/plugins/runtime/gateway/listener_generator.go

```diff
@@ -1,21 +1,243 @@
 	"github.com/kumahq/kuma/pkg/xds/generator"
 )
 
-// ListenerGenerator generates Kuma gateway listeners. Not implemented.
+const DefaultConnectionBuffer = 32 * 1024
+const DefaultRequestHeadersTimeoutMsec = 500 * time.Millisecond
+const DefaultStreamIdleTimeoutMsec = 5 * time.Second
+const DefaultConcurrentStreams = 100
+const DefaultInitialStreamWindowSize = 64 * 1024
+const DefaultInitialConnectionWindowSize = 1024 * 1024
+const DefaultIdleTimeout = 5 * time.Minute
+
+type gatewayPolicyAdaptor struct {
+	*core_mesh.GatewayResource
+}
+
+func (g gatewayPolicyAdaptor) Selectors() []*mesh_proto.Selector {
+	return g.Sources()
+}
+
+func findMatchingGateway(m manager.ReadOnlyResourceManager, dp *core_mesh.DataplaneResource) *core_mesh.GatewayResource {
+	gatewayList := &core_mesh.GatewayResourceList{}
+
+	if err := m.List(context.Background(), gatewayList, store.ListByMesh(dp.Meta.GetMesh())); err != nil {
+		return nil
+	}
+
+	candidates := make([]policy.DataplanePolicy, len(gatewayList.Items))
+	for i, gw := range gatewayList.Items {
+		candidates[i] = gatewayPolicyAdaptor{gw}
+	}
+
+	if p := policy.SelectDataplanePolicy(dp, candidates); p != nil {
+		return p.(gatewayPolicyAdaptor).GatewayResource
+	}
+
+	return nil
+}
+
+var _ policy.DataplanePolicy = gatewayPolicyAdaptor{}
+
+// ListenerGenerator generates Kuma gateway listeners.
 type ListenerGenerator struct {
 	Resources manager.ReadOnlyResourceManager
 }
 
 var _ generator.ResourceGenerator = ListenerGenerator{}
 
-func (ListenerGenerator) Generate(ctx xds_context.Context, proxy *xds.Proxy) (*xds.ResourceSet, error) {
-	return nil, errors.New("not implemented")
+func (l ListenerGenerator) Generate(_ xds_context.Context, proxy *core_xds.Proxy) (*core_xds.ResourceSet, error) {
+	gw := findMatchingGateway(l.Resources, proxy.Dataplane)
+	if gw == nil {
+		log.V(1).Info("no matching gateway for dataplane",
+			"name", proxy.Dataplane.Meta.GetName(),
+			"mesh", proxy.Dataplane.Meta.GetMesh(),
+		)
+
+		return nil, nil
+	}
+
+	log.V(1).Info(fmt.Sprintf("matched gateway %q to dataplane %q",
+		gw.Meta.GetName(), proxy.Dataplane.Meta.GetName()))
+
+	// Multiple listener specifications can have the same port. If
+	// they are compatible, then we can collapse those specifications
+	// down to a single listener.
+	collapsed := map[uint32][]*mesh_proto.Gateway_Listener{}
+	for _, ep := range gw.Spec.GetConf().GetListeners() {
+		collapsed[ep.GetPort()] = append(collapsed[ep.GetPort()], ep)
+	}
+
+	// A Gateway is a single service across all listeners.
+	service := proxy.Dataplane.Spec.GetIdentifyingService()
+
+	resources := core_xds.NewResourceSet()
+
+	// Generate a listener resource for each port.
+	for port, listeners := range collapsed {
+		protocol := listeners[0].GetProtocol()
+		address := proxy.Dataplane.Spec.GetNetworking().Address
+
+		log.V(1).Info("generating listener",
+			"address", address,
+			"port", port,
+			"protocol", protocol,
+		)
```

Comment:
so far we were sort of hesitant to log even into V(1) in generators. This can produce many logs if something in the cluster is changing a lot and you've got a lot of instances of gateway.

I think we can keep it for now since it's applied only for gateways, but please keep an eye on it.

Reply:
Yeh we can review the logging when we complete the feature :)

Reply:
Ok, please make sure we don't forget to revisit it